keps/sig-api-machinery/4008-crd-ratcheting/README.md

```diff
@@ -0,0 +1,1272 @@
+by the rule, but that is OK it just makes the rule more conservative.
+
+From this we can conclude that a given validation rule is completely dependent 
+upon a subset of the information contained in the field to which it is attached.
+If this subset of data did not change over the update, then the failure of the
+rule was definitely due to pre-existing data.
+
+For beta and beyond we will have an investigation and discussion of a more
+complicated semantic.
+
+For now we use DeepEqual to check equality since it is conservative. A semantic
+check may be employed if it can be done cleanly and performantly.
+
+There are a few cases which, to support ratcheting may require a less strict
+ratcheting rule:
+
+1. `maxItems` on an array - should we consider the "size" of the array to remain
+unchanged if elements within it change without adding/removing anything? 
+The current rule does not allow for this to be racheted.
+2. `maxProperties` - similarly to above
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+At a high level, CR validation in kubernetes has four components:
+1. `kube-openapi`-based `Validate` using `*spec.Schema`
+2. StructuralSchema-based Meta, TypeMeta, ListSet/Map validation
+3. CEL Validators using a CEL schema interface type
+4. Embedded Objects
+
+Each of these validation paths will need to have their `ValidateUpdate(old, new)`
+codepath modified to:
+
+1. Take an option to enable the feature
+2. Ignore errors which occur both on the `old` and `new` objects
+3. Return ignored errors in a separate list
+
+### `kube-openapi` changes
+
+The `kube-openapi` schema validator validates the standard OpenAPI/JSONSchema
+elements of the CRD schema. This is basically everything except CEL rules,
+and listset/map types.
+
+The current SchemaValidator will be wrapped with a `RatchetingSchemaValidator`
+which compares old/new values and ignores all errors on its subpath if they are
+equal.
+
+Supporting validation code within kube-openapi which spawn a new `SchemaValidator`
+will need to be refactored to find the old value and use `RatchetingSchemaValidator`
+when `RatchetingSchemaValidator.ValidateUpdate` is in its callstack.
+
+anyOf, allOf, not, allOf sections of logic will remain using the normal 
+`SchemaValidator`.
+
+#### Nested Value Validations 
+
+OpenAPI-schemas support nested validations:
+- not
+- oneOf
+- anyOf
+- allOf
+
+
+To keep the semantics simple and implementation clear, none of these constructs
+nor any of their subschemas will be ratcheted. Our intution is that in the wild
+CRD authors do not make much use of them, and when they do the logic is 
+complicated and references multiple fields.
```

Comment:
Not to be clarified in this KEP, but I believe your kube-openapi PoC actually ratcheting at the root of a schema (e.g. schema of a property). If not/oneOf/anyOf/allOf happen to be used on that node in the schema, the upper ratcheting will basically override whatever the not/oneOf/anyOf/allOf validator is doing. So the key here is that you skip using `RatchetingSchemaValidator` for anything inside those junctor validators.